### PR TITLE
MF-932 - User API keys

### LINKED ISSUE #932: User API keys
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.
No.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
In #778 we decided to introduce API keys. Introduction of Users API keys is the first step in that direction. Users API keys are similar to JWTs: the User can issue these keys with the configurable expiration time. These keys can later be used as API keys in custom, third-party services to act on behalf of the User.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
This is a must-have feature.

### PR DESCRIPTION
**Pull request by dborovcanin**
### What does this do?
This pull request introduces the API keys service. In the first version, only User keys will be supported.

### Which issue(s) does this PR fix/relate to?
This pull request resolves #932.

### List any changes that modify/break current functionality
Creating User token (JWT) is extracted from the Users service in favor of newly introduced API keys service.

### Have you included tests for your changes?
Yes.

### Did you document any new/modified functionality?
Yes.

### CODE REVIEW

File: auth/postgres/init.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	SSLCert     string
+	SSLKey      string
+	SSLRootCert string
+}
+
+// Connect creates a connection to the PostgreSQL instance and applies any
+// unapplied database migrations. A non-nil error is returned to indicate
+// failure.
+func Connect(cfg Config) (*sqlx.DB, error) {
+	url := fmt.Sprintf("host=%s port=%s user=%s dbname=%s password=%s sslmode=%s sslcert=%s sslkey=%s sslrootcert=%s", cfg.Host, cfg.Port, cfg.User, cfg.Name, cfg.Pass, cfg.SSLMode, cfg.SSLCert, cfg.SSLKey, cfg.SSLRootCert)
```
Comment (manuio) [inline auth/postgres/init.go]: you can maybe brake this in multiple lines

File: users/api/endpoint_test.go [hunk 2]
```diff
@@ -350,7 +351,11 @@ func TestPasswordChange(t *testing.T) {
 	ts := newServer(svc)
 	defer ts.Close()
 	client := ts.Client()
-	j := jwt.New("secret")
+	// j := jwt.New("secret")
```
Comment (manuio) [inline users/api/endpoint_test.go]: Dead code

**Review by manuio — CHANGES_REQUESTED** [review]

File: auth/postgres/database.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/opentracing/opentracing-go"
+)
+
+var _ Database = (*database)(nil)
+
+type database struct {
+	db *sqlx.DB
+}
+
+// Database provides a database interface
```
Comment (nmarcetic) [inline auth/postgres/database.go]: Is it possible to move this to tracing package ? I would expect to see something else in database package, file named database.go

File: auth/postgres/init.go (hunk 1 above)
Comment (nmarcetic) [reply auth/postgres/init.go]: Maybe using [string literals](https://golang.org/ref/spec#String_literals)

File: auth/postgres/key.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	q := `DELETE FROM keys WHERE issuer = $1 AND id = $2`
+
+	if _, err := kr.db.ExecContext(ctx, q, issuer, id); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type dbKey struct {
```
Comment (nmarcetic) [inline auth/postgres/key.go]: I would raither store only revoked keys in DB, not valid keys.

File: auth/api/logging.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+		}
+		lm.logger.Info(fmt.Sprintf("%s without errors.", message))
+	}(time.Now())
+
+	return lm.svc.Retrieve(ctx, owner, id)
+}
+
+func (lm *loggingMiddleware) Identify(ctx context.Context, key string) (id string, err error) {
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method identify for took %s to complete", time.Since(begin))
```
Comment (drasko) [inline auth/api/logging.go]: Fix message

File: auth/key.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	"errors"
+	"time"
+)
+
+var (
+	// ErrInvalidKeyIssuedAt indicates that the Key is being used before it's issued.
+	ErrInvalidKeyIssuedAt = errors.New("invalid issue time")
+
+	// ErrKeyExpired indicates that the Key is expired.
+	ErrKeyExpired = errors.New("use o expired key")
```
Comment (drasko) [inline auth/key.go]: Fix message

File: auth/postgres/key.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	// required for DB access
```
Comment (drasko) [inline auth/postgres/key.go]: Remove useless comment

File: auth/postgres/key.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	errDuplicate = "unique_violation"
+	errInvalid   = "invalid_text_representation"
+)
+
+type keyRepository struct {
+	db Database
+}
+
+// New instantiates a PostgreSQL implementation of key
+// repository.
```
Comment (drasko) [inline auth/postgres/key.go]: No need breaking this comment in two lines, especially that lines of code below are longer

**Review by drasko — CHANGES_REQUESTED** [review]

File: .env [hunk 9]
```diff
@@ -26,10 +26,20 @@ MF_JAEGER_CONFIGS=5778
 MF_JAEGER_URL=jaeger:6831
 
 ## Core Services
+
+### Authn
```
Comment (drasko) [inline .env]: `AuthN`

File: auth.proto [hunk 10]
```diff
@@ -13,12 +13,13 @@ service ThingsService {
     rpc Identify(Token) returns (ThingID) {}
 }
 
-service UsersService {
+service AuthnService {
```
Comment (drasko) [inline auth.proto]: `AuthN`

File: authn/api/grpc/client.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+	kitot "github.com/go-kit/kit/tracing/opentracing"
+	kitgrpc "github.com/go-kit/kit/transport/grpc"
+	opentracing "github.com/opentracing/opentracing-go"
+
+	"github.com/mainflux/mainflux"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+var _ mainflux.AuthnServiceClient = (*grpcClient)(nil)
```
Comment (drasko) [inline authn/api/grpc/client.go]: `AuthN`

File: authn/api/grpc/server.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	kitgrpc "github.com/go-kit/kit/transport/grpc"
+	mainflux "github.com/mainflux/mainflux"
+	"github.com/mainflux/mainflux/authn"
+	opentracing "github.com/opentracing/opentracing-go"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var _ mainflux.AuthnServiceServer = (*grpcServer)(nil)
```
Comment (drasko) [inline authn/api/grpc/server.go]: `AuthN`

File: cmd/users/main.go [hunk 13]
```diff
@@ -230,25 +244,45 @@ func connectToDB(dbConfig postgres.Config, logger logger.Logger) *sqlx.DB {
 		logger.Error(fmt.Sprintf("Failed to connect to postgres: %s", err))
 		os.Exit(1)
 	}
+
 	return db
 }
 
-func newService(db *sqlx.DB, tracer opentracing.Tracer, c config, logger logger.Logger) users.Service {
+func connectToAuthn(cfg config, tracer opentracing.Tracer, logger logger.Logger) (mainflux.AuthnServiceClient, func() error) {
```
Comment (drasko) [inline cmd/users/main.go]: `AuthN`

File: cmd/users/main.go [hunk 14]
```diff
@@ -179,23 +194,22 @@ func loadConfig() config {
...
+		logLevel:      mainflux.Env(envLogLevel, defLogLevel),
+		dbConfig:      dbConfig,
+		authnHTTPPort: mainflux.Env(envAuthnHTTPPort, defAuthnHTTPPort),
+		authnGRPCPort: mainflux.Env(envAuthnGRPCPort, defAuthnGRPCPort),
+		authnURL:      mainflux.Env(envAuthnURL, defAuthnURL),
+		authnTimeout:  time.Duration(timeout) * time.Second,
+		authnTLS:      tls,
+		emailConf:     emailConf,
+		httpPort:      mainflux.Env(envHTTPPort, defHTTPPort),
+		// grpcPort:      mainflux.Env(envGRPCPort, defGRPCPort),
```
Comment (drasko) [inline cmd/users/main.go]: Dead code?

**Review by drasko — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
@dusanb94 just minor remarks, and I let you decide if `AuthN` should be used intead of `Authn`. Also, please stay aligned to a #931 (and vice versa)

File: authn/service.go [hunk 15]
```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package authn
+
+import (
+	"context"
+	"errors"
```
Comment (blokovi) [inline authn/service.go:8]: You can use `"github.com/mainflux/mainflux/errors"` here

File: authn/postgres/init.go [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+					`CREATE TABLE IF NOT EXISTS keys (
+						id          UUID NOT NULL,
+						type        SMALLINT,
+						issuer      VARCHAR(254) NOT NULL,
+						issued_at   TIMESTAMP NOT NULL,
+						expires_at  TIMESTAMP,
+						PRIMARY KEY (id, issuer)
+					)`,
+				},
+				Down: []string{"DROP TABLE IS EXISTS keys"},
```
Comment (mteodor) [inline authn/postgres/init.go]: IF?

File: .env [hunk 17]
```diff
@@ -26,10 +26,20 @@ MF_JAEGER_CONFIGS=5778
...
+MF_AUTHN_GRPC_PORT=8181
+MF_AUTHN_DB_PORT=5432
+MF_AUTHN_DB_USER=mainflux
+MF_AUTHN_DB_PASS=mainflux
+MF_AUTHN_DB=auth
+MF_AUTHN_SECRET=secret
+
 ### Users
 MF_USERS_LOG_LEVEL=debug
 MF_USERS_HTTP_PORT=8180
```
Comment (mteodor) [inline .env:42]: MF_USERS_GRPC_PORT remove all references if not needed

File: authn/api/doc.go [hunk 18]
```diff
@@ -0,0 +1,5 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+// Package api contains implementation of Authn service HTTP API.
```
Comment (manuio) [inline authn/api/doc.go]: AuthN

File: users/service.go [hunk 19]
```diff
@@ -146,13 +139,12 @@ func (svc usersService) UserInfo(ctx context.Context, token string) (User, error
...
 	}, nil
-
 }
 
 func (svc usersService) UpdateUser(ctx context.Context, token string, u User) errors.Error {
-	email, err := svc.idp.Identity(token)
-	if err != nil {
-		return ErrUnauthorizedAccess
+	email, err := svc.identify(ctx, token)
+	if err != nil || email != u.Email {
```
Comment (manuio) [inline users/service.go]: This test is not necessary and probably will add a bug. I recently fixed this in the master: https://github.com/mainflux/mainflux/pull/959/files

**Review by manuio — CHANGES_REQUESTED** [review]

File: authn/service.go (hunk 15 above)
Comment (dborovcanin) [reply authn/service.go:8]: @blokovi Yes, that's the plan. There is a separate issue for that: #970, and I'll resolve it ASAP.

File: authn/service.go [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+type service struct {
+	keys KeyRepository
+	idp  IdentityProvider
+	t    Tokenizer
+}
+
+// New instantiates the auth service implementation.
+func New(keys KeyRepository, idp IdentityProvider, tokenizer Tokenizer) Service {
+	return &service{
+		t:    tokenizer,
```
Comment (manuio) [inline authn/service.go]: This could be `tokenizer:    t,`

File: authn/api/http/endpoint_test.go [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+	uk := issueRequest{Type: authn.UserKey, Duration: time.Hour}
+
+	cases := []struct {
+		desc   string
+		req    string
+		ct     string
+		token  string
+		status int
+	}{
+		{"issue login key", toJSON(lk), contentType, "", http.StatusCreated},
```
Comment (anovakovic01) [inline authn/api/http/endpoint_test.go]: Use named fields for test cases.

File: authn/postgres/key.go [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+		return authn.Key{}, err
+	}
+
+	return toKey(key), nil
+}
+
+func (kr repo) Remove(ctx context.Context, issuer, id string) error {
+	q := `DELETE FROM keys WHERE issuer = $1 AND id = $2`
+
+	if _, err := kr.db.ExecContext(ctx, q, issuer, id); err != nil {
```
Comment (anovakovic01) [inline authn/postgres/key.go]: Why aren't you using `NamedExecContext` here?

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: mqtt/aedes/package-lock.json [hunk 23]
```diff
@@ -13,8 +13,8 @@
       "resolved": "https://registry.npmjs.org/@grpc/proto-loader/-/proto-loader-0.5.2.tgz",
       "integrity": "sha512-eBKD/FPxQoY1x6QONW2nBd54QUEyzcFP9FenujmoeDPy1rutVSHki1s/wR68F6O1QfCNDx+ayBH1O2CVNMzyyw==",
       "requires": {
-        "lodash.camelcase": "4.3.0",
-        "protobufjs": "6.8.8"
+        "lodash.camelcase": "^4.3.0",
```
Comment (nmarcetic) [inline mqtt/aedes/package-lock.json]: Not sure if those changes are made by mistake, but there is a lot of version changes. Can you please check ?

**Review by nmarcetic — DISMISSED** [review]

File: auth/postgres/key.go (hunk 4 above)
Comment (nmarcetic) [reply auth/postgres/key.go]: You don't actually store the token :) I just figured out,  so all good.

**Review by manuio — DISMISSED** [review]
LGTM

**Review by drasko — DISMISSED** [review]
LGTM

File: authn/uuid/idp.go [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"github.com/gofrs/uuid"
+	"github.com/mainflux/mainflux/authn"
+)
+
+var _ authn.IdentityProvider = (*uuidIdentityProvider)(nil)
+
+type uuidIdentityProvider struct{}
+
+// New instantiates a UUID identity provider.
```
[the remark this answers is not in the document]
Comment (dborovcanin) [reply authn/uuid/idp.go:16]: I don't think that we have real benefits from it, but it's aligned with other services, so I'd keep it for now. I agree that we should probably move it to `pkg` or `internal`. We can do the same for the mocks.

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Review by nmarcetic — APPROVED** [review]

**Pull request merged**